rust/lance/src/dataset.rs

```diff
@@ -1065,6 +1067,40 @@ impl Dataset {
         }
     }
 
+    /// Get a stream of batches based on iterator of slices of row numbers.
+    ///
+    /// This is an experimental API. It may change at any time.
+    pub fn take_scan(
+        &self,
+        row_slices: Box<dyn Iterator<Item = Result<(u64, u64)>> + Send>,
+        projection: Arc<Schema>,
+        batch_readahead: usize,
+    ) -> DatasetRecordBatchStream {
+        let arrow_schema = Arc::new(projection.as_ref().into());
+        let dataset = Arc::new(self.clone());
+        let batch_stream = futures::stream::iter(row_slices)
+            .map(move |res| {
+                let dataset = dataset.clone();
+                let projection = projection.clone();
+                let fut = async move {
+                    let (start, end) =
+                        res.map_err(|err| DataFusionError::External(Box::new(err)))?;
+                    let row_pos: Vec<u64> = (start..end).collect();
+                    dataset
+                        .take(&row_pos, projection.as_ref())
+                        .await
+                        .map_err(|err| DataFusionError::External(Box::new(err)))
+                };
+                async move { tokio::spawn(fut).await.unwrap() }
```

Comment:
```suggestion
                async move {
                    let (start, end) =
                        res.map_err(|err| DataFusionError::External(Box::new(err)))?;
                    let row_pos: Vec<u64> = (start..end).collect();
                    dataset
                        .take(&row_pos, projection.as_ref())
                        .await
                        .map_err(|err| DataFusionError::External(Box::new(err)))
                }
```
Can you just return `fut` instead of wrapping it in the second `async move` block?

Reply:
THe `fut` is so I can pass to `tokio::spawn`, so these can run over multiple threads. I did some tracing and found this improved performance quite a bit.

I can move the `tokio::spawn` inside the async block so there is only one of them though.

Reply:
Actually, this turned out to be necessary to keep as is, for the sake of the lifetimes. `tokio::task::spawn` needs statics lifetimes for the future, so we need a separate async block that captures the `dataset`, `projection`, and `row_pos`.